### PR TITLE
Modal Support

### PR DESCRIPTION
**Pull request by etrotta**
Issue #74  for context

### CODE REVIEW

File: flask_discord_interactions/models/modal.py
Comment (Breq16) [inline flask_discord_interactions/models/modal.py:10]: This class will need to be added to the documentation. You've written nice docstrings (which is most of the work honestly) but the class name needs to be added to a file in the `docs` directory for Sphinx to pick up on it. Ideally I'd like to have a brief written overview of using modals, some short code examples, and then the autodoc.

You can see a good example in the autocomplete docs, file is [here](https://github.com/Breq16/flask-discord-interactions/blob/664e70908a597c344eac2792422896de71b71194/docs/autocomplete.rst) and output is [here](https://flask-discord-interactions.readthedocs.io/en/latest/autocomplete.html).

If you're familiar with Sphinx/reStructuredText for docs or willing to learn, I'd encourage you to give them a try; if not, I'm happy to write up the page for this.

File: flask_discord_interactions/enums/__init__.py
Comment (Breq16) [inline flask_discord_interactions/enums/__init__.py:4]: I'm a bit hesitant to introduce an entirely new sub-package just for a single 7-line enum, but I do see the value of this going forward. I suppose the cleaner code is probably worth the complexity.

**Review by Breq16 — CHANGES_REQUESTED** [review]
Thanks for opening this PR! Most of the diff look good, but there are a few changes I'd like before I'm ready to merge this.

File: flask_discord_interactions/enums/__init__.py
Comment (etrotta) [reply flask_discord_interactions/enums/__init__.py:4]: We probably should move it to somewhere in the models, but I'm not sure which file would be the most appropriate (or create a new file in there). That said, if you're fine with the folder, it might be OK to keep. I don't have much experience designing APIs or long projects though.

File: flask_discord_interactions/tests/test_command.py
Comment (etrotta) [reply flask_discord_interactions/tests/test_command.py:27]: I think that some of them might have done automatically by my editor because of some plugins I have installed. Will revert though

File: flask_discord_interactions/discord.py
Comment (etrotta) [inline flask_discord_interactions/discord.py:624]: This one can stay? Comparing with `== None` is generally not a good idea, despite probably not mattering here.

File: docs/modal.rst
Comment (etrotta) [inline docs/modal.rst:65]: A bit of a draft at best - feel free to change it (please)
I'm not sure if it should mention which fields TextInput accepts or not, but I left it there

File: flask_discord_interactions/models/modal.py
Comment (etrotta) [reply flask_discord_interactions/models/modal.py:10]: I started a bit of a draft, but yeah, I might leave it up to you.

**Comment by etrotta** [comment]
I might have done a few too many small commits, but wanted to address each 'topic' in a separate one.
There are still a few things to iron out, specially now that I remembered about buttons and others being allowed to return modals, but I'll leave the rest for another day (probably tomorrow).
Thanks for the in depth code review by the way!

File: flask_discord_interactions/discord.py
Comment (Breq16) [reply flask_discord_interactions/discord.py:630]: Yep, this is fine.

File: flask_discord_interactions/enums/__init__.py
Comment (Breq16) [reply flask_discord_interactions/enums/__init__.py:4]: We could do something like `models/response.py`, and have a superclass (say, `Response`) of both `Message` and `Modal`. It would be nice, semantically, for every command function to return some type of `Response` object. That said, I don't think they have a ton of logic in common.

**Comment by Breq16** [comment]
Of course! Thanks for being open to suggestions. And don't worry about the number of commits--personally I think it's nice to have a detailed record of each minor change. No rush to get this done, I've got some midterms coming up so I might not be able to review your changes immediately anyway.

In regards to the custom ID handlers (buttons etc), it'd be nice if we didn't have the `isinstance(..., Modal)` logic in multiple places. Maybe we could replace `Message.from_return_value` with a method somewhere else. (Although I don't think it could be in the same place as the `ResponseType` enum due to circular input issues.)

File: docs/modal.rst
Comment (Breq16) [reply docs/modal.rst:65]: Honestly this looks good! re: TextInput fields, I think the autodoc will take care of that.

File: docs/modal.rst
Comment (Breq16) [inline docs/modal.rst:57]: This is pretty verbose, I can see now why you wanted to flatten the `Context.components` to remove the action rows. I do think we can still make this cleaner without resorting to that though. Maybe a method like `Context.get_component("...")` which would take in a custom ID string and return the first component in the tree that matches it?

Comment (Breq16) [reply docs/modal.rst:57]: Looks good!

File: flask_discord_interactions/enums/__init__.py
Comment (Breq16) [reply flask_discord_interactions/enums/__init__.py:4]: I haven't really thought of a better way to do this, so the enum directory can stay for now.

**Comment by Breq16** [comment]
Finally got the time to take another look at this. I pushed a quick change to the documentation page.

I also changed how components are loaded; the incoming component tree is made up of `Component` objects instead of plain dicts. This should make user code less prone to typo-ing and make the syntax a bit nicer.

As far as I'm concerned, this is pretty much ready to merge at this point. Unless you have any objections, I'll probably merge this sometime tonight.

**Pull request merged**
